torch/csrc/utils/torch_dispatch_mode.h

```diff
@@ -8,16 +8,15 @@ namespace torch_dispatch_mode {
 struct StashTorchDispatchModeGuard {
  public:
   StashTorchDispatchModeGuard() {
-    saved_ = c10::impl::TorchDispatchModeTLS::get_state();
-    c10::impl::TorchDispatchModeTLS::set_state(nullptr);
+    c10::impl::TorchDispatchModeTLS::swap_mode(saved_mode_);
```

Comment:
Guards will just save the C++ mode, not the whole stack. This is because it's mostly used by the python_args_parser which just turns off the current C++ and runs its `torch_dispatch`. We need the mode to remain because the stack mode's `torch_dispatch` relies on popping an element from the top of the stack to call its `torch_dispatch`

Reply:
Let me suggest something more radical: why don't we just pop the mode off the stack at the call sites for StashTorchDispatchModeGuard. Since we moved the mode stack to C++, we should also move the pushing popping, and morally what we want to do before the torch dispatch call is to just push the mode that we are about to execute off the stack